### PR TITLE
Rings new

### PR DESCRIPTION
**Pull request by fgrunewald**
pending pysmiles 2.0

### CODE REVIEW

File: cgsmiles/pysmiles_utils.py [hunk 1]
```diff
@@ -30,27 +27,23 @@ def rebuild_h_atoms(mol_graph, keep_bonding=False):
...
-                hcount -= len(mol_graph.nodes[node]['bonding'])
-
-            mol_graph.nodes[node]['hcount'] = hcount
-            if ele == "H":
-                mol_graph.nodes[node]['single_h_frag'] = True
+        if mol_graph.nodes[node].get('aromatic', False):
+            mol_graph.nodes[node]['hcount'] = 0
+
+        if mol_graph.nodes[node].get('bonding', False) and  \
+        mol_graph.nodes[node].get('ele,emt', '*') == "H":
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py]:
```suggestion
        mol_graph.nodes[node].get('element', '*') == "H":
```

File: cgsmiles/pysmiles_utils.py [hunk 2]
```diff
@@ -30,27 +27,23 @@ def rebuild_h_atoms(mol_graph, keep_bonding=False):
...
-            # in this case we only rebuild hydrogen atoms that are not
-            # replaced by bonding operators.
-            if keep_bonding:
-                hcount -= len(mol_graph.nodes[node]['bonding'])
-
-            mol_graph.nodes[node]['hcount'] = hcount
-            if ele == "H":
-                mol_graph.nodes[node]['single_h_frag'] = True
+        if mol_graph.nodes[node].get('aromatic', False):
+            mol_graph.nodes[node]['hcount'] = 0
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py:31]: Why 0?

File: cgsmiles/pysmiles_utils.py [hunk 3]
```diff
@@ -30,27 +27,23 @@ def rebuild_h_atoms(mol_graph, keep_bonding=False):
...
-            bonds = math.ceil(sum([mol_graph.edges[(node, neigh)]['order'] for neigh in\
-                               mol_graph.neighbors(node)]))
-            charge = mol_graph.nodes[node].get('charge', 0)
-            hcount = pysmiles.smiles_helper._valence(mol_graph, node, minimum=0) -\
-                     bonds +\
-                     charge
-            # in this case we only rebuild hydrogen atoms that are not
-            # replaced by bonding operators.
-            if keep_bonding:
-                hcount -= len(mol_graph.nodes[node]['bonding'])
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py:46]: This seems rather important, and it disappeared?

File: cgsmiles/resolve.py [hunk 4]
```diff
@@ -189,6 +182,10 @@ def edges_from_bonding_descrpt(self):
                 # unless they are specifically annotated
                 order = int(bonding[0][-1])
                 self.molecule.add_edge(edge[0], edge[1], bonding=bonding, order=order)
+                if self.all_atom:
+                    for edge_node in edge:
+                        if self.molecule.nodes[edge_node]['element'] != 'H':
+                            self.molecule.nodes[edge_node]['hcount'] -= 1
```
Comment (pckroon) [inline cgsmiles/resolve.py:188]:
```suggestion
                            self.molecule.nodes[edge_node]['hcount'] -= order
```
?

**Review by pckroon — CHANGES_REQUESTED** [review]

File: cgsmiles/pysmiles_utils.py (hunk 2 above)
Comment (fgrunewald) [reply cgsmiles/pysmiles_utils.py:31]: the accounting of hydrogen is not correct after stitching them together. Therefore we set them all to zero / undefined which is equivalent to what pysmiles would expect when reading a string for the first time. Then we kekulize

File: cgsmiles/resolve.py (hunk 4 above)
Comment (fgrunewald) [reply cgsmiles/resolve.py:188]: this works is the short answer; having the order there does not presumably due to aromatic bond orders

File: cgsmiles/pysmiles_utils.py (hunk 3 above)
Comment (fgrunewald) [reply cgsmiles/pysmiles_utils.py:46]: because this is actually rather tricky thing. How do you deal with aromatic bonding connectors?

**Comment by pckroon** [comment]
Hmmn, about the aromatic bond orders. I'd say you round the aromatic bond order down (to 1), and after, if your atom is aromatic you subtract another hydrogen.

Let me have a look at the code

File: .github/workflows/python-app.yml [hunk 5]
```diff
@@ -16,7 +16,7 @@ jobs:
     
     strategy:
       matrix:
-       py_version: ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12"]
+       py_version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (pckroon) [inline .github/workflows/python-app.yml:19]: This change should also be reflected in `setup.cfg`

**Review by pckroon — CHANGES_REQUESTED** [review]

**Review by pckroon — APPROVED** [review]

**Pull request merged**